### PR TITLE
fix global performance node 16

### PR DESCRIPTION
**Pull request by benjamingr**
#### Purpose (TL;DR) - mandatory

Fixes `performance` on Node 16 which has a global `performance` but not a global `Performance` 

cc @fatso83 @SimenB @itayperry

### CODE REVIEW

File: src/fake-timers-src.js [hunk 1]
```diff
@@ -1,6 +1,13 @@
 "use strict";
 
-const globalObject = require("@sinonjs/commons").global;
+// Node.js stores certain things like the performance object on globalThis
+// but not on its `global` object which means we prefer it to `global` in
+// node versions that support it - diverging from the @sinonjs/commons behavior
+/* global globalThis */
```
Comment (fatso83) [inline src/fake-timers-src.js:6]: This is all fine and dandy, but is there a reason we should not include this change in `@sinonjs/commons`?

**Review by fatso83 — COMMENTED** [review]
Big <3 for this. Just a question before merging.

File: src/fake-timers-src.js (hunk 1 above)
Comment (benjamingr) [reply src/fake-timers-src.js:6]: Well, it changes the meaning of what `global` is and I am not sure that applies to other sinon projects. Let me take another look at the code.

Comment (benjamingr) [reply src/fake-timers-src.js:6]: I'm trying to figure out why performance won't be on `global`

File: test/fake-timers-test.js [hunk 2]
```diff
@@ -3775,11 +3778,26 @@ describe("FakeTimers", function () {
                 delete Performance.prototype.someFunc3;
             });
 
+            it("should mock performance on Node 16+", function () {
+                // node 16+ has a performance object but not a global constructor
+                if (typeof performance === "undefined") {
+                    this.skip();
```
Comment (SimenB) [inline test/fake-timers-test.js:3784]: shouldn't there be a `return` after the `skip`s so we don't do further works that's ignored?

File: src/fake-timers-src.js [hunk 3]
```diff
@@ -1672,7 +1685,7 @@ function withGlobal(_global) {
                                 : NOOP;
                     }
                 });
-            } else if (config.toFake.includes("performance")) {
+            } else if ((config.toFake || []).includes("performance")) {
```
Comment (SimenB) [inline src/fake-timers-src.js:1688]: seems unrelated? when is it missing, and how hasn't that been an issue before?

Comment (benjamingr) [reply src/fake-timers-src.js:1688]: This has been an issue with Node 16 causing sinon to crash. It's here now in case something in the future reaches this branch

File: test/fake-timers-test.js (hunk 2 above)
Comment (benjamingr) [reply test/fake-timers-test.js:3784]: I'm... not sure should there?

File: src/fake-timers-src.js (hunk 3 above)
Comment (itayperry) [reply src/fake-timers-src.js:1688]: > how hasn't that been an issue before?

It was in the latest merged PR #400 and it works on Node 14-15 but not on Node 16  (it was my mistake)

see https://github.com/sinonjs/fake-timers/issues/394#issuecomment-1013758365

Comment (SimenB) [reply src/fake-timers-src.js:1688]: cool. can we add node 16 (and 17) to CI with this PR?

File: test/fake-timers-test.js (hunk 2 above)
Comment (benjamingr) [reply test/fake-timers-test.js:3784]: No other places in the code do this - so I guessing "no" but I'm just cargo-culting here

File: src/fake-timers-src.js (hunk 3 above)
Comment (benjamingr) [reply src/fake-timers-src.js:1688]: Sure, added Node 16

Comment (benjamingr) [reply src/fake-timers-src.js:1688]: Let me remove 10 which is EoL and has stopped receiving security updates and add Node 17 and a test for something that exists there

Comment (benjamingr) [reply src/fake-timers-src.js:1688]: Node 17 has unrelated failures, let me try and fix those

Comment (benjamingr) [reply src/fake-timers-src.js:1688]: I'll also add a bunch of todos for stuff that doesn't work atm

Comment (fatso83) [reply src/fake-timers-src.js:1688]: Not totally surehow to get that coverage thing going.

Comment (SimenB) [reply src/fake-timers-src.js:1688]: removing 10 is a breaking change, fwiw

Comment (fatso83) [reply src/fake-timers-src.js:1688]: @SimenB Nothing is broken and what we actually run the tests on is not our compatibility test. We have an official policy for all Sinon projects: https://github.com/sinonjs/sinon/blob/master/COMPATIBILITY.md#runtimes which basically says "All supported LTS versions". Once 10 no longer is supported, then it's off the table. No need to bump major version numbers. Not that it matters all that much :smiley:

Comment (SimenB) [reply src/fake-timers-src.js:1688]: Right, I just saw "Let me remove 10 which is EoL". If that doesn't happen then my comment can be ignored 👍 

---

It's still a breaking change to remove remove support for a runtime, which warrants a new major version regardless of what some markdown file says (unless it also says "we do not follow semver, and a new major means incompatible API change" or some such).

Anyways, this probably got a bit theoretical and not related to the PR at hand 😀

Comment (SimenB) [reply src/fake-timers-src.js:1688]: > (I don't think supporting EoL versions of Node makes sense since they have bugs/security issues that are never going to be fixed)

I don't disagree with that statement. I'm not saying `@sinonjs/fake-timers` should keep supporting EOL versions of Node indefinitely (or even that this PR shouldn't remove support for it, although that seems orthogonal to fixing support for the current LTS release), but _if_ (when) support is dropped, that's a breaking change which should trigger a new major version.

**Comment by benjamingr** [comment]
Let's merge and open another PR to discuss versioning?

**Comment by fatso83** [comment]
I am merging and ignoring the codecov failures due to dropped coverage, but it would be good if that was analyzed somehow so that we understand why. I am guessing the various "if x is present, run this, else that" is only covered if the tests are run in several runtimes (which it is), but are the codecov results just one of these runs or the aggregated result? If the former, then it's hard to do something about it, and if the latter, it is weird that we get a drop.

File: test/fake-timers-test.js (hunk 2 above)
Comment (fatso83) [reply test/fake-timers-test.js:3784]: Either `return` or use a `if/else`. Ref the [Mocha docs:](https://mochajs.org/#:~:text=Best%20practice%3A%20To%20avoid%20confusion%2C%20do%20not%20execute%20further%20instructions%20in%20a%20test%20or%20hook%20after%20calling%20this.skip().)
> _Best practice:_ To avoid confusion, do not execute further instructions in a test or hook after calling this.skip().

I can fix this.

**Comment by jakub-g** [comment]
FWIW for future readers: using fake-timers@9.1.2 with nodejs 16.3.0:

```js
jest.useFakeTimers({
    now: ...
});
```

I have 

```
TypeError: Cannot assign to read only property 'performance' of object '[object global]'
```

Solution: upgrade to nodejs 16.5.0+ or use

```js
jest.useFakeTimers({
    now: ...
    doNotFake: ['performance'],
});
```
